resources/type-system/flow-analysis.md

```diff
@@ -0,0 +1,713 @@
+    - and `captured` is false
+    - and `S` is the current type of `x` in `VM`
+    - and `T <: S` and not `S <: T`
+    - and `T` is a type of interest for `x` in `tested`
+
+  - We say that a variable `x` is demotable via assignment of an expression of
+    type `T` given variable model `VM` if
+    - `VM = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - and `captured` is false
+    - and declared::promoted contains a type `S` such that `T` is `S` or `T` is
+      **NonNull(`S`)**.
+
+Definitions:
+
+- `assign(x, E, M)` where `x` is a local variable, `E` is an expression of
+  inferred type `T`, and `M = FlowModel(r, VI)` is the flow model for `E` is
+  defined to be `FlowModel(r, VI[x -> VM])` where:
+    - `VI(x) = VariableModel(declared, promoted, tested, assigned, unassigned, captured)`
+    - if `captured` is true then:
+      - `VM = VariableModel(declared, promoted, tested, true, false, captured)`.
+    - otherwise if `x` is promotable via initialization given `VM` then
+      - `VM = VariableModel(declared, [T], tested, true, false, captured)`.
+    - otherwise if `x` is promotable via assignment of `E` given `VM`
+      - `VM = VariableModel(declared, T::promoted, tested, true, false, captured)`.
+    - otherwise if `x` is demotable via assignment of `E` given `VM`
+      - `VM = VariableModel(declared, demoted, tested, true, false, captured)`.
+      - where `previous` is the suffix of `promoted` starting with the first type
+        `S` such that `T <: S`, and:
+        - if `S`is nullable and if `T <: Q` where `Q` is **NonNull(`S`)** then
+          `demoted` is `Q::previous`
+        - otherwise `demoted` is `previous`
+
+- `promote(E, T, M)` where `E` is an expression, `T` is a type which it may be
+  promoted to, and `M = FlowModel(r, VI)` is the flow model in which to promote,
+  is defined as follows:
+  - If `E` is not a promotion target, then `M`
+  - If `E` is a promotion target `x`, then
+    - Let `VM = VariableModel(declared, promoted, tested, assigned, unassigned,
+      captured)` be the variable model for `x` in `VI`
+    - If `x` is not promotable via type test to `T` given `VM`, then return `M`
+    - Else
+      - Let `S` be the current type of `x` in `VM`
+      - If `T <: S` then let `T1` = `T`
+      - Else if `S` is `X extends R` then let `T1` = `X & T`
+      - Else If `S` is `X & R` then let `T1` = `X & T`
+      - Else `x` is not promotable (shouldn't happen since we checked above)
+      - Let `VM2 = VariableModel(declared, T1::promoted, T1::tested, assigned,
+      unassigned, captured)`
+      - Return `FlowModel(r, VI[x -> VM2])`
+- `promoteToNonNull(E, M)` where `E` is an expression and `M` is a flow model is
+  defined to be `promote(E, T, M)` where `T0` is the type of `E`, and `T` is
+  **NonNull(`T0`)**.
+- `factor(T, S)` where `T` and `S` are types defines the "remainder" of `T` when
```

Comment:
The implementation currently doesn't have any of this logic.  I'm tracking in https://github.com/dart-lang/sdk/issues/40014, however I must admit that I'm questioning whether this feature is worth the implementation effort.  The only non-contrived example I can come up with where this is useful is that it would cause `if (x is! Null)` to promote `x` in the same way that `if (x != null)` would.  Do you have a more compelling example?

Reply:
I'll put discussion in the issue.